### PR TITLE
feat: gnosis node support, part 1

### PR DESCRIPTION
**Pull request by lmyslinski**
This PR adds Gnosis nodes to our backend. It doesn't include the indexer (blockbook) as well as the unchained API, which will come in a separate PR.

Gnosis is set up on Nethermind + Lighthouse instead of Geth, which results in significantly faster sync times.

### CODE REVIEW

**Comment by lmyslinski** [comment]
Manually verified that both nodes use the correct data directory:
```
49M     ./var
132G    ./data
166M    ./nethermind
132G    .
132G    total
```

**Comment by lmyslinski** [comment]
Both nodes running with readiness probes detecting sync state:
Screenshot from 2023-05-08 17-41-22

File: node/coinstacks/gnosis/daemon/init.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
+
+set -e
+
+start_coin_bg() {
+  ./Nethermind.Runner --config gnosis --datadir /data/gnosis --JsonRpc.Host=0.0.0.0 --JsonRpc.Port=8545 --JsonRpc.WebSocketsPort=8546 --JsonRpc.JwtSecretFile=/jwt.hex --Network.ActivePeersMaxCount=100 --HealthChecks.Enabled=true --Metrics.CountersEnabled=true
```
[the remark this answers is not in the document]
Comment (kaladinlight) [reply node/coinstacks/gnosis/daemon/init.sh:6]: yeah cli flags are easier to manage/update with our setup than mounting an extra config file

File: node/coinstacks/gnosis/api/src/app.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import { service } from './controller'
+import { RegisterRoutes } from './routes'
+
+const PORT = process.env.PORT ?? 3000
+const INDEXER_WS_URL = process.env.INDEXER_WS_URL
+
+if (!INDEXER_WS_URL) throw new Error('INDEXER_WS_URL env var not set')
+
+export const logger = new Logger({
+  namespace: ['unchained', 'coinstacks', 'ethereum', 'api'],
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/api/src/app.ts:25]:
```suggestion
  namespace: ['unchained', 'coinstacks', 'gnosis', 'api'],
```

File: node/coinstacks/gnosis/api/src/app.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  level: process.env.LOG_LEVEL,
+})
+
+const app = express()
+
+app.use(json())
+app.use(urlencoded({ extended: true }))
+app.use(cors())
+
+app.get('/health', async (_, res) => res.json({ status: 'up', asset: 'ethereum', connections: wsServer.clients.size }))
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/api/src/app.ts:35]:
```suggestion
app.get('/health', async (_, res) => res.json({ status: 'up', asset: 'gnosis', connections: wsServer.clients.size }))
```

File: node/coinstacks/gnosis/api/src/app.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+const options: swaggerUi.SwaggerUiOptions = {
+  customCss: '.swagger-ui .topbar { display: none }',
+  customSiteTitle: 'ShapeShift Ethereum API Docs',
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/api/src/app.ts:39]:
```suggestion
  customSiteTitle: 'ShapeShift Gnosis API Docs',
```

File: node/coinstacks/gnosis/api/src/controller.ts [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+  BaseAPI,
+  BaseInfo,
+  InternalServerError,
+  SendTxBody,
+  ValidationError,
+} from '../../../common/api/src' // unable to import models from a module with tsoa
+import { API, Account, GasFees, Tx, TxHistory, GasEstimate } from '../../../common/api/src/evm' // unable to import models from a module with tsoa
+import { Service } from '../../../common/api/src/evm/service'
+
+const ETHERSCAN_API_KEY = process.env.ETHERSCAN_API_KEY
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/api/src/controller.ts:16]: will need to create a gnosisscan api key and add to circle env vars

File: node/coinstacks/gnosis/api/src/controller.ts [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+const RPC_URL = process.env.RPC_URL
+
+if (!ETHERSCAN_API_KEY) throw new Error('ETHERSCAN_API_KEY env var not set')
+if (!INDEXER_URL) throw new Error('INDEXER_URL env var not set')
+if (!INDEXER_WS_URL) throw new Error('INDEXER_WS_URL env var not set')
+if (!NETWORK) throw new Error('NETWORK env var not set')
+if (!RPC_URL) throw new Error('RPC_URL env var not set')
+
+export const logger = new Logger({
+  namespace: ['unchained', 'coinstacks', 'ethereum', 'api'],
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/api/src/controller.ts:29]:
```suggestion
  namespace: ['unchained', 'coinstacks', 'gnosis', 'api'],
```

File: node/coinstacks/gnosis/api/src/controller.ts [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+  level: process.env.LOG_LEVEL,
+})
+
+const blockbook = new Blockbook({ httpURL: INDEXER_URL, wsURL: INDEXER_WS_URL })
+const provider = new ethers.providers.JsonRpcProvider(RPC_URL)
+
+export const service = new Service({
+  blockbook,
+  explorerApiKey: ETHERSCAN_API_KEY,
+  explorerApiUrl: 'https://api.etherscan.io/api',
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/api/src/controller.ts:39]:
```suggestion
  explorerApiUrl: 'https://api.gnosisscan.io/api',
```

File: node/coinstacks/gnosis/api/src/controller.ts [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+  provider,
+  logger,
+  rpcUrl: RPC_URL,
+})
+
+@Route('api/v1')
+@Tags('v1')
+export class Ethereum extends Controller implements BaseAPI, API {
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/api/src/controller.ts:47]:
```suggestion
export class Gnosis extends Controller implements BaseAPI, API {
```

File: node/coinstacks/gnosis/daemon/init.sh (hunk 1 above)
Comment (kaladinlight) [inline node/coinstacks/gnosis/daemon/init.sh:6]: we need the debug module for tracing with current logic: 
`--JsonRpc.EnabledModules=eth,net,web3,debug,subscribe,trace,txpool,health,rpc`

also let's remove the `ActivePeersMaxCount` it is listed as deprecated and no need to mess with peer count unless we are experienceing sync issues

File: node/coinstacks/gnosis/daemon/init.sh [hunk 9]
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
+
+set -e
+
+start_coin_bg() {
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/daemon/init.sh:5]: i've been using `start()` and `stop()` on more recent init scripts, doesn't matter too much, but just to make note

File: node/coinstacks/gnosis/indexer/config.json [hunk 10]
```diff
@@ -0,0 +1,21 @@
+{
+  "fiat_rates": "coingecko",
+  "fiat_rates_params": "{\"url\": \"https://api.coingecko.com/api/v3\", \"coin\": \"gnosis\",\"platformIdentifier\": \"gnosis\",\"platformVsCurrency\": \"gno\",\"periodSeconds\": 900}",
+  "fiat_rates_vs_currencies": "AED,ARS,AUD,BDT,BHD,BMD,BRL,CAD,CHF,CLP,CNY,CZK,DKK,EUR,GBP,HKD,HUF,IDR,ILS,INR,JPY,KRW,KWD,LKR,MMK,MXN,MYR,NGN,NOK,NZD,PHP,PKR,PLN,RUB,SAR,SEK,SGD,THB,TRY,TWD,UAH,USD,VEF,VND,ZAR,BTC,ETH",
+  "mempoolTxTimeoutHours": 24,
+  "queryBackendOnMempoolResync": false,
+  "coin_name": "Gnosis",
+  "coin_shortcut": "GNO",
+  "coin_label": "Gnosis",
+  "rpc_url": "ws://localhost:8545",
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/indexer/config.json:10]: i was also seeing some path routing in the node config `--help` with separate port have you confirmed `:8546` connects to websocket as expected?
```suggestion
  "rpc_url": "ws://localhost:8546",
```

File: node/coinstacks/gnosis/pulumi/Pulumi.yaml [hunk 11]
```diff
@@ -0,0 +1,3 @@
+name: gnosis
+runtime: nodejs
+description: coin stack indexer, ingestion and interface
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/pulumi/Pulumi.yaml:3]: some eth things are a little out of date, been just using `coin stack`
```suggestion
description: coin stack
```

File: node/coinstacks/gnosis/pulumi/index.ts [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+          ...service,
+          ports: {
+            'daemon-http': { port: 8545 },
+            'daemon-ws': { port: 8546 },
+            'daemon-beacon': { port: 8551, ingressRoute: false },
+          },
+          configMapData: {
+            'jwt.hex': readFileSync('../daemon/jwt.hex').toString(),
+          },
+          readinessEndpoint: '/health',
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/pulumi/index.ts:25]: wish more nodes exposed this :)

File: node/coinstacks/gnosis/pulumi/index.ts [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+          ...service,
+          command: [
+            '/usr/local/bin/lighthouse',
+            'beacon_node',
+            '--network=gnosis',
+            '--disable-upnp',
+            '--datadir=/data',
+            '--port=9000',
+            '--http',
+            '--http-address=0.0.0.0',
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/pulumi/index.ts:39]: we can leave the beacon addresses default, we aren't exposing publicly

File: node/coinstacks/gnosis/pulumi/index.ts [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+            '--http-port=5052',
+            '--target-peers=50',
+            '--execution-endpoint=http://localhost:8551',
+            '--execution-jwt=/jwt.hex',
+            '--debug-level=info',
+            '--validator-monitor-auto',
+            '--subscribe-all-subnets',
+            '--import-all-attestations',
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/pulumi/index.ts:47]: seems like validator related config, not sure we need/care

File: node/coinstacks/gnosis/pulumi/index.ts [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+      case 'daemon-beacon':
+        return {
+          ...service,
+          command: [
+            '/usr/local/bin/lighthouse',
+            'beacon_node',
+            '--network=gnosis',
+            '--disable-upnp',
+            '--datadir=/data',
+            '--port=9000',
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/pulumi/index.ts:37]: not as concerned about explicit default port for the beacon chain if we just want to use the defaults and save some lines

File: node/coinstacks/gnosis/sample.env [hunk 16]
```diff
@@ -0,0 +1,6 @@
+# ENVIRONMENT VARIABLES
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/sample.env:1]: will want to add `ETHERSCAN_API_KEY` as `# SECRET ENVIRONMENT VARIABLES`

File: pulumi/src/docker.ts [hunk 17]
```diff
@@ -59,7 +59,9 @@ export const buildAndPushImage = async (args: BuildAndPushImageArgs): Promise<vo
 
   dockerBuildArgs.push(args.context)
 
-  execSync(`${envs} docker build --platform linux/amd64 ${dockerBuildArgs.join(' ')}`, { stdio: 'inherit' })
+  execSync(`${envs} docker build --platform linux/amd64 -o type=docker ${dockerBuildArgs.join(' ')}`, {
```
Comment (kaladinlight) [inline pulumi/src/docker.ts:62]: just curious, are you manually changing these or do you have some formatting config set in your IDE doing this? have been noticing a lot of these the last couple prs

File: pulumi/src/statefulService.ts [hunk 18]
```diff
@@ -59,6 +59,18 @@ export function createCoinService(args: CoinServiceArgs, assetName: string, snap
         ...(args.memoryRequest && { memory: args.memoryRequest }),
       },
     },
+    ...(args.readinessEndpoint && {
```
Comment (kaladinlight) [inline pulumi/src/statefulService.ts:62]: I think we should leverage the existing `readinessProbe` config and set on the container  if `(!readiness && { ...args.readinessProbe })`, we can update `init`, `liveness`, and `readiness` to `initFile`, `livenessFile` and `readinessFile` to clarify as well

File: node/coinstacks/gnosis/docker-compose.yml [hunk 19]
```diff
@@ -0,0 +1,14 @@
+version: '3.6'
+
+services:
+  api:
+    image: unchained-local-node
+    env_file: .env
+    labels:
+      - 'traefik.enable=true'
+      - 'traefik.http.routers.gnosis-api.rule=Host(`api.gnosis.localhost`)'
```
Comment (kaladinlight) [inline node/coinstacks/gnosis/docker-compose.yml:9]: will need to add the docker network to `unchained/docker-compose.yml`

File: node/coinstacks/gnosis/pulumi/index.ts (hunk 15 above)
Comment (lmyslinski) [reply node/coinstacks/gnosis/pulumi/index.ts:37]: All of these are straight up taken from Gnosis docs, sure we can prune them

File: pulumi/src/docker.ts (hunk 17 above)
Comment (lmyslinski) [reply pulumi/src/docker.ts:62]: Ugh, docker is pain. Turns out that if you install just the docker engine, not docker desktop (as I just did on my linux box) it doesn't generate a docker image automatically, so I've had to add it explicitly.

File: node/coinstacks/gnosis/daemon/init.sh (hunk 1 above)
Comment (lmyslinski) [reply node/coinstacks/gnosis/daemon/init.sh:6]: Ad 2. Used the wrong prop but here's the reasoning for bumping it up:

```
if you have a good connection and a reasonable machine then setting --Network.MaxActivePeers to 256 (see NetworkConfig secion in configuration) should give much better fast sync times (we use 256 peers to get 5 hours syncs)
```
https://docs.nethermind.io/nethermind/ethereum-client/sync-modes

File: node/coinstacks/gnosis/pulumi/index.ts (hunk 12 above)
Comment (lmyslinski) [reply node/coinstacks/gnosis/pulumi/index.ts:25]: Given how fast nethermind is, might be moving all geth nodes to it down the line

File: node/coinstacks/gnosis/indexer/config.json (hunk 10 above)
Comment (lmyslinski) [reply node/coinstacks/gnosis/indexer/config.json]: Screenshot from 2023-05-09 14-19-34

File: node/coinstacks/gnosis/pulumi/index.ts (hunk 12 above)
Comment (kaladinlight) [reply node/coinstacks/gnosis/pulumi/index.ts:25]: certainly new integrations if supported. once geth is synced, it is pretty bullet proof. would be interesting to see if rpc request performance differs a lot between the two. that would be a good reason to upgrade for sure

File: pulumi/src/docker.ts (hunk 17 above)
Comment (kaladinlight) [reply pulumi/src/docker.ts:62]: oh, i didn't even notice the type addition, i was just asking about the formatting of the objects on multiple lines vs single line.

are you using docker buildx? I am on linux without docker desktop and haven't seen this issue. it looks like this bundles docker archive images and I'm not sure we want this. feels like more of a local environment issue that needs to be resolved.

File: monitoring/src/config.ts [hunk 20]
```diff
@@ -26,7 +26,7 @@ export const getConfig = async (): Promise<LoopConfig> => {
   const kubeconfig = (await stackReference.getOutputValue('kubeconfig')) as string
   const namespaces = (await stackReference.getOutputValue('namespaces')) as Array<string>
   const defaultNamespace = (await stackReference.getOutputValue('defaultNamespace')) as string
-  const domain = (await stackReference.getOutputValue('rootDomainName')) as string
+  const domain = process.env.ADDITIONAL_ROOT_DOMAIN_NAME as string
```
Comment (kaladinlight) [inline monitoring/src/config.ts:29]: revert

File: pulumi/src/docker.ts (hunk 17 above)
Comment (lmyslinski) [reply pulumi/src/docker.ts:62]: kk will revert for now

**Pull request merged**